Add /quest tolocation command

fingoltin opened the pull request:
Still to do: Add persistence with serialization.

fingoltin committed:
Finish /quest tolocation

fingoltin committed:
Add checks to QuestLocation

If the player already has the quest, don't prompt.
Checks to see if the QuestLocation intersects with another, and if it
does, don't create it.

fingoltin committed:
Add user feedback to /quest tolocation

HaedHutner commented on src/main/java/com/atherys/quests/commands/quest/AttachQuestToLocationCommand.java:
```diff
@@ -1,30 +1,48 @@
-public class AttachQuestToLocationCommand implements CommandExecutor {
+@Aliases("tolocation")
+@Description("Attaches a quest to a location.")
+public class AttachQuestToLocationCommand implements CommandExecutor, ParameterizedCommand {
```
There's no need to implement both CommandExecutor and ParameterizedCommand, as ParameterizedCommand already extends CommandExecutor

HaedHutner commented on src/main/java/com/atherys/quests/listeners/EntityListener.java:
```diff
@@ -26,6 +29,17 @@ public void onEntityInteract(InteractEntityEvent.Secondary.MainHand event, @Root
 
+    @Listener
+    public void onPlayerMove(MoveEntityEvent e, @Root Player player){
+        if(LocationManager.getInstance().getByLocation(e.getFromTransform().getLocation()).isPresent()) return;
```
Check whether or not the position of the player is the same between transforms to avoid unnecessary work being done by the plugin

HaedHutner commented on src/main/java/com/atherys/quests/managers/LocationManager.java:
```diff
@@ -14,12 +16,45 @@
 
-    protected LocationManager() {
+    private static LocationManager instance = new LocationManager();
+    private static Gson gson = GsonUtils.getGson();
```
Ideally, getGson would be called during initialization of the LocationManager. Having it as a static field introduces possible ndefined behaviors.

HaedHutner commented on src/main/java/com/atherys/quests/managers/LocationManager.java:
```diff
@@ -14,12 +16,45 @@
+       saveAll(this.getCache().values());
+    }
+
+    public boolean addLocationQuest(Location<World> location, String questId, double radius){
```
addQuestLocation may be a better name here?

HaedHutner commented on src/main/java/com/atherys/quests/managers/LocationManager.java:
```diff
@@ -37,17 +72,29 @@ protected LocationManager() {
 
         private String questId;
 
+        private boolean overlaps(QuestLocation questLocation){
```
use Vector3d#distanceSquared() for better performance, sqrt is a heavier operation than squaring the radius you wish to compare against

fingoltin committed:
Performance improvements

fingoltin committed:
Add persistence, but still issues

Gson isn't properly creating the objective objects,
and the amount of QuestLocations doubles every time the server
launches.

fingoltin committed:
Merge branch 'master' of https://github.com/Atherys-Horizons/AtherysQuests

fingoltin committed:
Fix quest deserialization

fingoltin committed:
Refactor, fix PartyMsg to QuestMsg

fingoltin committed:
Update gitignore

fingoltin committed:
Merge branch 'refactor'

fingoltin committed:
Remove unused imports

fingoltin committed:
Consistency

LJNIC commented:
This PR fixed #1 and added QuestLocation's. But my mongo database is doubling the amount of QuestLocation's every shutdown, and I don't know why.

HaedHutner commented on src/main/java/com/atherys/quests/managers/LocationManager.java:
```diff
@@ -14,18 +17,66 @@
-        return Optional.empty();
+        Document document = new Document();
+
+        Document location = DbUtils.Serialize.location(questLocation.location);
```
If you have registered a ConfigurateAdapter<Location> within the Gson instance, there is no need for DbUtils.

HaedHutner commented on src/main/java/com/atherys/quests/managers/LocationManager.java:
```diff
@@ -14,18 +17,66 @@
+        document.append("location", location);
+
+        document.append("questId", questLocation.questId);
+        document.append("radius", questLocation.radius);
```
Please access these fields using getters. Also, create getters for them. Don't need to make setters, a QuestLocation can be Immutable, that would be just fine, but accessing them directly like this is quite ugly and not very safe.

fingoltin committed:
Use Gson instead of DbUtils, add DummyQuest to Adapter
